---
title: "Introduction to the VCFComparison package"
author:
- Le Zhang lzhan140@ucr.edu 
- Lei Yu lyu062@ucr.edu
date: "`r Sys.Date()`"
bibliography: references.bib  
link-citations: yes
output:
  html_document:
    toc: true
    toc_float: true
    
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction
### Structure variation

Structure variants (SVs) are large genomic alterations which typically larger than 50 bp. Nowdays, multiple SV callers can detect the structure variation such as DELLY, Manta, GRIDSS, LUMPY etc @rausch2012delly, @chen2016manta, @cameron2017gridss, @layer2014lumpy. 

A common challenge of SV detection is the large false positive rate. Even each caller can provide the filter parameter to help users to filter the SV calls, we have concerns about the accuracy of the result from only one caller. Therefore, we need to apply the SV calling through multiple callers about combine the result to detect the most consistent calls among callers. 

VCFComparison is a R package to help users combine the result from different caller and unify the calls then perform visualization on the detected SVs. 

## 1. Read the files.
### 1.1 Sample map file

Usually, in a structure variation detection project, researchers apply multiple callers in same cohorts. Therefore, users need to provide sample mapping data frame to indecate the sample ID and the directory of different callers on different sample. Here we use two callers and two patients as an example. 

**VCFComparison works based on the individual .vcf file from each caller.**

We define the sample mapping data frame as follow: 

| Sample ID | Caller 1 | Caller 2 |
|----------------|-----------------------------------------------------------------------|-----------------------------
| patient_1 | '/user/caller1/1.vcf' | '/user/caller2/1.vcf'| 
| patient_2 | '/user/caller1/2.vcf' | '/user/caller2/2.vcf'| 

```{r eval= F}
sample_map <- data.frame(patient_ID = c("patient_1", "patient_2"), 
                         DELLY = c("/rhome/lyu062/shared/coh_SV/DELLY/output/tumor_vs_normal/filtered/post_filtered/M02.vcf", "/rhome/lyu062/shared/coh_SV/DELLY/output/tumor_vs_normal/filtered/post_filtered/M03.vcf"), 
                         Manta = c("/rhome/lyu062/shared/coh_SV/Manta/T_vs_N_somatic_SV/converted_INV/sample_2.vcf", "/rhome/lyu062/shared/coh_SV/Manta/T_vs_N_somatic_SV/converted_INV/sample_3.vcf"))

```

### 1.2 Read the vcf files into R. 

Based on the sample mapping file, we can read the vcf files into R use the function Read_VCFs. 


```{r eval = FALSE}
VCF_list <- Read_VCFs(sample_map)
View(VCF_list)
```

### 1.3 Subtract the useful information from the VCF_list

There are three fields in a vcf file including meta, info and geno: 

**meta** the head lines in the vcf. meta-information is included after the ## string and must be key=value pairs.
Example: 

```{r eval = FALSE}
##fileformat=VCFv4.2
##fileDate=20090805
##source=myImputationProgramV3.1
##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta
##contig=<ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species="Homo sapiens",taxonomy=x>
##phasing=partial
##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">
##INFO=<ID=AA,Number=1,Type=String,Description="Ancestral Allele">
##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129">
##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership">
##FILTER=<ID=q10,Description="Quality below 10">
##FILTER=<ID=s50,Description="Less than 50% of samples have data">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality">
```

**fix** is a table consisted by 6 columns: CHROM POS ID REF ALT QUAL FILTER INFO
Example: 
```{r eval = FALSE}
xxxxx
```

**gt** notes the genotype information for of each samples appears in the vcf. 
Example

```{r eval = F}
 xxxx
```

In the research, users might not need to use the whole information from a vcf file. Therefore, one can use the 

## 1.3 Filter the vcf files
The hard filter step is based on the parameter ("PASS" and "Precise") on each callers. Users can set the "PASS", "Precise" or "Both" to filter the SVs. 

```{r eval = FALSE}


```

## 2. Merging the callers results of samples
Once we performed the  

```{r eval = FALSE}


```

## References

<div id="refs"></div>
